File: example-codes/blackjack-and-a-deck-of-cards/game.py
from .hand import  *
from .player import *
from .deck_shoe import *


def get_bet_from_UI():
    pass


def get_user_action():
    pass


class Game:
    def __init__(self, player, dealer):
        self.__player = player
        self.__dealer = dealer
        self.__MAX_NUM_OF_DECKS = 3
        self.__shoe = Shoe(self.__MAX_NUM_OF_DECKS)

    def play_action(self, action, hand):
        switcher = {
            "hit": self.hit(hand),
            "split": self.split(hand),
            "stand pat": None,  # do nothing
            "stand": self.stand()
        }
        switcher.get(action, 'Invalid move')

    def hit(self, hand):
        self.__hand.add_card(self.__shoe.deal_card())

    def stand(self):
        dealer_score = self.__dealer.get_total_score()
        player_score = self.__player.get_total_score()
        hands = self.__player.get_hands()
        for hand in hands:
            best_score = hand.resolve_score()
            if player_score == 21:
                # blackjack, pay 3: 2 of the bet
                None
            elif player_score > dealer_score:
                # pay player equal to the bet
                None
            elif player_score < dealer_score:
                # collect the bet from the player
                None
            else:  # tie
                # bet goes back to player
                None

    def split(self, hand):
        cards = hand.get_cards()
        self.__player.add_hand(Hand(cards[0], self.__shoe.deal_card()))
        self.__player.add_hand(Hand(cards[1], self.__shoe.deal_card()))
        self.__player.remove_hand(hand)

    def start(self):
        self.__player.place_bet(get_bet_from_UI())

        player_hand = Hand(self.__shoe.deal_card(), self.__shoe.deal_card())
        self.__player.add_to_hand(player_hand)

        dealer_hand = Hand(self.__shoe.deal_card(), self.__shoe.deal_card())
        self.__dealer.add_to_hand(dealer_hand)

        while True:
            hands = self.__player.get_hands()
            for hand in hands:
                action = get_user_action(hand)
                self.play_action(action, hand)
                if action.equals("stand"):
                    break


def main():
    player = Player()
    dealer = Dealer()
    game = Game(player, dealer)
    game.start()

File: example-codes/blackjack-and-a-deck-of-cards/java/Game.java
package blackjack;

import java.util.List;
import java.util.Scanner;

public class Game {
    private Player player;
    private Dealer dealer;
    private static final int MAX_NUM_OF_DECKS = 3;
    private Shoe shoe;

    public Game(Player player, Dealer dealer) {
        this.player = player;
        this.dealer = dealer;
        this.shoe = new Shoe(MAX_NUM_OF_DECKS);
    }

    public void playAction(String action, Hand hand) {
        switch (action.toLowerCase()) {
            case "hit":
                hit(hand);
                break;
            case "split":
                split(hand);
                break;
            case "stand pat":
                // do nothing
                break;
            case "stand":
                stand();
                break;
            default:
                System.out.println("Invalid move");
        }
    }

    public void hit(Hand hand) {
        hand.addCard(shoe.dealCard());
    }

    public void stand() {
        int dealerScore = dealer.getTotalScore();
        int playerScore = player.getTotalScore();
        List<Hand> hands = player.getHands();
        for (Hand hand : hands) {
            int bestScore = hand.resolveScore();
            if (bestScore == 21) {
                // blackjack, pay 3:2 of the bet
                System.out.println("Blackjack! Player wins 3:2");
            } else if (bestScore > dealerScore) {
                // pay player equal to the bet
                System.out.println("Player wins!");
            } else if (bestScore < dealerScore) {
                // collect the bet from the player
                System.out.println("Dealer wins!");
            } else { // tie
                // bet goes back to player
                System.out.println("Push! Tie game.");
            }
        }
    }

    public void split(Hand hand) {
        List<BlackjackCard> cards = hand.getCards();
        player.addHand(new Hand(cards.get(0), shoe.dealCard()));
        player.addHand(new Hand(cards.get(1), shoe.dealCard()));
        player.removeHand(hand);
    }

    public void start() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter your bet: ");
        double bet = scanner.nextDouble();
        player.placeBet(bet);

        Hand playerHand = new Hand(shoe.dealCard(), shoe.dealCard());
        player.addHand(playerHand);

        Hand dealerHand = new Hand(shoe.dealCard(), shoe.dealCard());
        dealer.addHand(dealerHand);

        while (true) {
            List<Hand> hands = player.getHands();
            for (Hand hand : hands) {
                String action = getUserAction(hand);
                playAction(action, hand);
                if (action.equals("stand")) {
                    break;
                }
            }
            break;
        }
        scanner.close();
    }

    private static double getBetFromUI() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter your bet: ");
        double bet = scanner.nextDouble();
        return bet;
    }

    private static String getUserAction(Hand hand) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Your hand score: " + hand.resolveScore());
        System.out.print("Enter action (hit/split/stand pat/stand): ");
        String action = scanner.nextLine();
        return action;
    }

    public static void main(String[] args) {
        Player player = new Player();
        Dealer dealer = new Dealer();
        Game game = new Game(player, dealer);
        game.start();
    }
}

